test/L2EthToken.spec.ts

```diff
@@ -0,0 +1,142 @@
+  });
+
+  it("withdrawWithMessage", async () => {
+    const iface = new ethers.utils.Interface([
+      "function finalizeEthWithdrawal(uint256, uint256, uint16, bytes, bytes32[])",
+    ]);
+    const selector = iface.getSighash("finalizeEthWithdrawal");
+    const amountToWidthdraw: BigNumber = ethers.utils.parseEther("1.0");
+    const additionalData: string = ethers.utils.defaultAbiCoder.encode(["string"], ["additional data"]);
+    const message: string = ethers.utils.solidityPack(
+      ["bytes4", "address", "uint256", "address", "bytes"],
+      [selector, l1Receiver.address, amountToWidthdraw, walletFrom.address, additionalData]
+    );
+
+    await setResult("L1Messenger", "sendToL1", [message], {
+      failure: false,
+      returnData: ethers.utils.defaultAbiCoder.encode(["bytes32"], [ethers.utils.keccak256(message)]),
+    });
+
+    const amountToWithdraw: BigNumber = ethers.utils.parseEther("1.0");
+
+    await expect(
+      l2EthToken.connect(walletFrom).withdrawWithMessage(l1Receiver.address, additionalData, {
+        value: amountToWithdraw,
+        gasLimit: 5000000,
+        gasPrice,
+      })
+    ).to.emit(l2EthToken, "WithdrawalWithMessage");
+  });
+
+  it("transfer successfully", async () => {
+    await l2EthToken.connect(bootloaderAccount).mint(walletFrom.address, ethers.utils.parseEther("100.0"));
+
+    const senderBalandeBeforeTransfer: BigNumber = await l2EthToken.balanceOf(walletFrom.address);
+    const recipientBalanceBeforeTransfer: BigNumber = await l2EthToken.balanceOf(walletTo.address);
+
+    const amountToTransfer = ethers.utils.parseEther("10.0");
+
+    await expect(
+      l2EthToken.connect(bootloaderAccount).transferFromTo(walletFrom.address, walletTo.address, amountToTransfer)
+    )
+      .to.emit(l2EthToken, "Transfer")
+      .withArgs(walletFrom.address, walletTo.address, amountToTransfer);
+
+    const senderBalanceAfterTransfer: BigNumber = await l2EthToken.balanceOf(walletFrom.address);
+    const recipientBalanceAfterTransfer: BigNumber = await l2EthToken.balanceOf(walletTo.address);
+    expect(senderBalanceAfterTransfer).to.be.eq(senderBalandeBeforeTransfer.sub(amountToTransfer));
+    expect(recipientBalanceAfterTransfer).to.be.eq(recipientBalanceBeforeTransfer.add(amountToTransfer));
+  });
+
+  it("no tranfser due to insufficient balance", async () => {
+    await l2EthToken.connect(bootloaderAccount).mint(walletFrom.address, ethers.utils.parseEther("5.0"));
+    const amountToTransfer: BigNumber = ethers.utils.parseEther("100000000000000000.0");
+
+    await expect(
+      l2EthToken.connect(bootloaderAccount).transferFromTo(walletFrom.address, walletTo.address, amountToTransfer)
+    ).to.be.rejectedWith("Transfer amount exceeds balance");
+  });
+
+  it("no transfer - require special access", async () => {
+    const provider: Provider = new Provider((hre.network.config as any).url);
+    const maliciousData = {
+      address: "0xdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef",
```

Comment:
I don't think this address corresponds to this private key, it's better to use the `maliciousWallet` address below.